File: java/src/main/java/net/hydrius/pacts/core/SchemaLoader.java
package net.hydrius.pacts.core;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;
import java.util.zip.ZipEntry;
import java.util.zip.ZipException;
import java.util.zip.ZipInputStream;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

/**
 * SchemaLoader class that loads schemas.
 * 
 * This class supports loading schemas from local and remote resources.
 */
public class SchemaLoader {

    private final ObjectMapper objectMapper;
    private final Map<String, JsonNode> cache;
    private static final Logger logger = Logger.getLogger(SchemaLoader.class.getName());

    private final String schemaRoot;
    private final String domain;
    private final String version;

    private static final int CONNECTION_TIMEOUT_MS = 15_000;
    private static final int READ_TIMEOUT_MS = 30_000;
    private static final long MAX_ENTRY_SIZE = 10 * 1024 * 1024; // 10 MB per ZIP entry

    /**
     * Creates a new SchemaLoader.
     *
     * @param schemaRoot the directory containing the schemas
     * @param domain     the domain of the schema
     * @param version    the version of the schema
     * @throws IllegalArgumentException if the schema root, domain, or version
     *                                  is null
     */
    public SchemaLoader(String schemaRoot, String domain, String version) {
        if (schemaRoot == null || domain == null || version == null) {
            throw new IllegalArgumentException("Schema root, domain, and version must be specified.");
        }

        this.objectMapper = new ObjectMapper();
        this.objectMapper.registerModule(new JavaTimeModule());
        this.cache = new HashMap<>();
        this.schemaRoot = schemaRoot;
        this.domain = domain;
        this.version = version;

        try {
            loadRemoteSchemas();
        } catch (IOException e) {
            logger.warning(() -> "Remote schema loading failed, falling back to local file system only: " + e.getMessage());
        }
    }

    /**
     * Loads a schema from cache or local resources.
     *
     * @param category The schema category (e.g., "player")
     * @param name     The schema name (e.g., "player_request")
     * @return The parsed JSON schema node or null if not found.
     */
    public JsonNode loadSchema(String category, String name) {
        String cacheKey = domain + "/" + version + "/" + category + "/" + name;

        if (cache.containsKey(cacheKey)) {
            return cache.get(cacheKey);
        }

        // Try to load from file system
        try {
            Path schemaPath = Paths.get(schemaRoot, domain, version, category, name + ".json");
            if (Files.exists(schemaPath)) {
                JsonNode schema = objectMapper.readTree(schemaPath.toFile());
                cache.put(cacheKey, schema);
                return schema;
            }
        } catch (IOException e) {
            logger.warning(() -> "Failed to load schema from file system: " + e.getMessage());
        }

        return null;
    }

    /**
     * Loads schemas from remote ZIP files specified in application.yml.
     * 
     * This method processes ZIP files in memory without writing to disk,
     * extracting all JSON schema files and storing them in an in-memory cache.
     * 
     * The loader supports multiple source URLs for redundancy, trying each one
     * in order until a successful load occurs.
     * 
     * @throws IOException if sources could not be read or found
     */
    private void loadRemoteSchemas() throws IOException {
        JsonNode settings = null;
        try (InputStream stream = getClass().getClassLoader().getResourceAsStream("application.yml")) {
            if (stream == null) {
                logger.warning("Application Settings not found");
            } else {
                ObjectMapper yamlMapper = new ObjectMapper(new YAMLFactory());
                JsonNode parsed = yamlMapper.readTree(stream);

                if (parsed == null || !parsed.has("sources")) {
                    logger.warning("application.yml is missing 'sources' key");
                } else {
                    JsonNode sourcesNode = parsed.get("sources");
                    if (!sourcesNode.isArray()) {
                        logger.warning("'sources' key in application.yml is not an array");
                    } else {
                        settings = objectMapper.createObjectNode();
                        ((ObjectNode) settings).set("sources", objectMapper.convertValue(sourcesNode, JsonNode.class));
                    }
                }
            }
        } catch (Exception e) {
            logger.warning(() -> "Failed to parse application settings: " + e.getMessage());
        }

        boolean sourcesLoaded = false;
        if (settings != null) {
            JsonNode sourcesNode = settings.get("sources");
            for (JsonNode node : sourcesNode) {
                String source = node.asText();
                HttpURLConnection connection = null;
                try {
                    logger.info(() -> "Attempting to load schemas from: " + source);
                    URL url = new URL(source);
                    connection = (HttpURLConnection) url.openConnection();
                    connection.setRequestMethod("GET");
                    connection.setConnectTimeout(CONNECTION_TIMEOUT_MS);
                    connection.setReadTimeout(READ_TIMEOUT_MS);

                    try (InputStream in = connection.getInputStream();
                         ZipInputStream zipIn = new ZipInputStream(in)) {

                        ZipEntry entry = zipIn.getNextEntry();
                        while (entry != null) {
                            if (!entry.isDirectory() && entry.getName().endsWith(".json")) {
                                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                                byte[] buffer = new byte[1024];
                                long totalRead = 0;
                                int len;
                                while ((len = zipIn.read(buffer)) > 0) {
                                    totalRead += len;
                                    if (totalRead > MAX_ENTRY_SIZE) {
                                        throw new IOException("ZIP entry exceeds maximum allowed size: " + entry.getName());
                                    }
                                    baos.write(buffer, 0, len);
                                }

                                byte[] content = baos.toByteArray();
                                JsonNode schema = objectMapper.readTree(content);

                                String entryPath = entry.getName();
                                int lastSlash = entryPath.lastIndexOf('/');
                                String fileName = lastSlash >= 0 ? entryPath.substring(lastSlash + 1) : entryPath;
                                String categoryName = lastSlash >= 0 ? entryPath.substring(0, lastSlash) : "";

                                String[] pathParts = categoryName.split("/");
                                if (pathParts.length >= 3) {
                                    String entryDomain = pathParts[pathParts.length - 3];
                                    String entryVersion = pathParts[pathParts.length - 2];
                                    String entryCategory = pathParts[pathParts.length - 1];
                                    String schemaName = fileName.substring(0, fileName.length() - 5);

                                    String cacheKey = entryDomain + "/" + entryVersion + "/" + entryCategory + "/"
                                            + schemaName;
                                    cache.put(cacheKey, schema);
                                    logger.info(() -> "Loaded schema into cache: " + cacheKey);
                                }
                            }

                            zipIn.closeEntry();
                            entry = zipIn.getNextEntry();
                        }
                    }

                    sourcesLoaded = true;
                    logger.info(() -> "Successfully loaded schemas from: " + source);
                    break;
                } catch (ZipException e) {
                    logger.warning(() -> "Failed to process ZIP from source: " + source + ", trying next source if available");
                } catch (IOException e) {
                    logger.warning(() -> "IO Exception while processing source: " + source + ", trying next source if available");
                } finally {
                    if (connection != null) {
                        connection.disconnect();
                    }
                }
            }
        }

        if (!sourcesLoaded) {
            throw new IOException("Sources could not be read or found to populate schemas.");
        }
    }

    /**
     * Clears all cached schemas.
     */
    public void clearCache() {
        cache.clear();
    }

    /**
     * Gets the schema root.
     *
     * @return the schema root
     */
    public String getSchemaRoot() {
        return schemaRoot;
    }

    /**
     * Gets the domain.
     *
     * @return the domain
     */
    public String getDomain() {
        return domain;
    }

    /**
     * Gets the version.
     *
     * @return the version
     */
    public String getVersion() {
        return version;
    }

    /**
     * Gets the parsed version.
     *
     * @return the parsed version
     */
    public int getParsedVersion() {
        return Integer.parseInt(version.replace("v", ""));
    }
}

File: rust/src/core/schema_loader.rs
use anyhow::Result;
use serde::Deserialize;
use serde_json::Value;
use std::collections::HashMap;
use std::io::Read;

use zip::read::ZipArchive;

#[derive(Deserialize)]
struct SourcesConfig {
    sources: Vec<String>,
}

/// SchemaLoader struct that loads schemas from remote sources.
/// 
/// This class loads schemas exclusively from remote ZIP files.
/// No local file system or embedded resources are used.
#[derive(Clone)]
pub struct SchemaLoader {
    schema_cache: HashMap<String, Value>,
    schema_root: String,
    domain: String,
    version: String,
}

impl SchemaLoader {
    /// Creates a new SchemaLoader and loads schemas from remote sources.
    ///
    /// # Arguments
    /// * `schema_root` - reserved for future local file system loading (must be non-empty)
    /// * `domain` - the domain of the schema
    /// * `version` - the version of the schema
    /// 
    /// # Panics
    /// Panics if the schema root, domain, or version is empty, or if remote schemas cannot be loaded
    pub fn new(schema_root: String, domain: String, version: String) -> Self {
        if schema_root.is_empty() || domain.is_empty() || version.is_empty() {
            panic!("Schema root, domain, and version must be specified.");
        }

        let mut loader = Self {
            schema_cache: HashMap::new(),
            schema_root,
            domain,
            version,
        };
        
        // Load remote schemas on initialization - must succeed
        loader.load_remote_schemas().expect("Failed to load remote schemas");
        
        loader
    }

    /// Loads a schema from cache by category and name.
    ///
    /// # Arguments
    /// * `category` - The schema category (e.g., "player")
    /// * `name` - The schema name (e.g., "player_request")
    ///
    /// # Returns
    /// The parsed JSON schema value
    ///
    /// # Panics
    /// Panics if the schema cannot be found in cache
    pub fn load_schema(&mut self, category: &str, name: &str) -> Value {
        let cache_key = format!("{}/{}/{}/{}", self.domain, self.version, category, name);

        // Only check cache - no fallbacks
        if let Some(schema) = self.schema_cache.get(&cache_key) {
            return schema.clone();
        }

        panic!(
            "Schema not found in cache: {}/{}/{}/{}",
            self.domain, self.version, category, name
        );
    }

    /// Loads schemas from remote ZIP files specified in sources.yaml.
    /// 
    /// This method processes ZIP files in memory without writing to disk,
    /// extracting all JSON schema files and storing them in an in-memory cache.
    /// 
    /// The loader supports multiple source URLs for redundancy, trying each one
    /// in order until a successful load occurs.
    /// 
    /// # Returns
    /// Result indicating success or failure
    fn load_remote_schemas(&mut self) -> Result<()> {
        // Read sources from configuration file
        let sources = self.load_sources_config()?;
        
        let mut sources_loaded = false;
        
        for source in sources {
            let cache_size_before = self.schema_cache.len();
            match self.load_schemas_from_zip_url(&source) {
                Ok(_) => {
                    if self.schema_cache.len() > cache_size_before {
                        eprintln!("Successfully loaded schemas from: {}", source);
                        sources_loaded = true;
                        break;
                    } else {
                        eprintln!("ZIP from {} contained no loadable schemas, trying next source", source);
                    }
                }
                Err(e) => {
                    eprintln!("Failed to load schemas from {}: {}", source, e);
                }
            }
        }
        
        if sources_loaded {
            Ok(())
        } else {
            Err(anyhow::anyhow!("Sources could not be read or found to populate schemas."))
        }
    }
    
    /// Loads sources configuration embedded at compile time from sources.yaml.
    /// 
    /// # Returns
    /// Result containing vector of source URLs
    fn load_sources_config(&self) -> Result<Vec<String>> {
        const SOURCES_YAML: &str = include_str!(concat!(env!("OUT_DIR"), "/sources.yaml"));

        let config: SourcesConfig = serde_yaml::from_str(SOURCES_YAML)
            .map_err(|e| anyhow::anyhow!("Failed to parse embedded sources.yaml: {}", e))?;

        if config.sources.is_empty() {
            return Err(anyhow::anyhow!("No sources defined in sources.yaml"));
        }

        Ok(config.sources)
    }
    
    /// Loads schemas from a ZIP file at the given URL.
    /// 
    /// # Arguments
    /// * `url` - The URL of the ZIP file to load schemas from
    /// 
    /// # Returns
    /// Result indicating success or failure
    fn load_schemas_from_zip_url(&mut self, url: &str) -> Result<()> {
        let agent: ureq::Agent = ureq::Agent::config_builder()
            .timeout_global(Some(std::time::Duration::from_secs(30)))
            .build()
            .into();

        let mut response = agent.get(url)
            .call()
            .map_err(|e| anyhow::anyhow!("HTTP request to {} failed: {}", url, e))?;

        const MAX_RESPONSE_SIZE: u64 = 50 * 1024 * 1024; // 50 MB total
        let mut bytes = Vec::new();
        response.body_mut().as_reader().take(MAX_RESPONSE_SIZE).read_to_end(&mut bytes)?;
            
        // Process the ZIP archive in memory
        let reader = std::io::Cursor::new(bytes);
        let mut zip = ZipArchive::new(reader)?;
        
        // Iterate through all entries in the ZIP file
        for i in 0..zip.len() {
            let mut entry = zip.by_index(i)?;
            
            // Process only JSON files
            if !entry.is_dir() && entry.name().ends_with(".json") {
                let entry_name = entry.name().to_string();
                
                // Read the entry content
                let mut content = String::new();
                if let Err(e) = entry.read_to_string(&mut content) {
                    eprintln!("Failed to read entry {} (index {}): {}", entry_name, i, e);
                    continue;
                }
                
                // Parse the JSON schema
                let schema: Value = match serde_json::from_str(&content) {
                    Ok(s) => s,
                    Err(e) => {
                        eprintln!("Failed to parse JSON for entry {} (index {}): {}", entry_name, i, e);
                        continue;
                    }
                };
                
                // Extract path information to create cache key
                let path_parts: Vec<&str> = entry_name.split('/').collect();
                
                // We expect paths like: Schemas-main/bees/v1/inventory/item.json
                if path_parts.len() >= 5 {
                    let entry_domain = path_parts[path_parts.len() - 4];
                    let entry_version = path_parts[path_parts.len() - 3];
                    let entry_category = path_parts[path_parts.len() - 2];
                    let file_name = path_parts[path_parts.len() - 1];
                    
                    let schema_name = file_name.trim_end_matches(".json");
                    
                    let cache_key = format!("{}/{}/{}/{}", entry_domain, entry_version, entry_category, schema_name);
                    self.schema_cache.insert(cache_key.clone(), schema);
                    eprintln!("Loaded schema into cache: {}", cache_key);
                } else {
                    eprintln!("Skipping JSON file with unexpected path structure (< 5 components): {}", entry_name);
                }
            }
        }
        
        Ok(())
    }

    /// Clears all cached schemas.
    pub fn clear_cache(&mut self) {
        self.schema_cache.clear();
    }

    /// Gets the schema root.
    ///
    /// # Returns
    /// the schema root
    pub fn get_schema_root(&self) -> &str {
        &self.schema_root
    }

    /// Gets the domain.
    ///
    /// # Returns
    /// the domain
    pub fn get_domain(&self) -> &str {
        &self.domain
    }

    /// Gets the version.
    ///
    /// # Returns
    /// the version
    pub fn get_version(&self) -> &str {
        &self.version
    }

    /// Gets the parsed version.
    ///
    /// # Returns
    /// the parsed version as an integer
    pub fn get_parsed_version(&self) -> i32 {
        self.version.replace("v", "").parse().unwrap_or(1)
    }
}
